aws_ad_accounts_sync.py

```diff
@@ -0,0 +1,283 @@
+    changed_attributes.append('signing certificate(s)')
+  if delete_inline_user_policies(iam, user_name):
+    changed_attributes.append('inline user policy attachment(s)')
+  if detach_managed_user_policies(iam, user_name):
+    changed_attributes.append('managed user policy attachment(s)')
+  iam.delete_user(UserName=user_name)
+  delete_message     = 'user: %s account: %s has been deleted after no activity for %s days.' % (user_name, account, aws_delete_grace_period)
+  attributes_message = ' These attributes were deleted: %s' % str(changed_attributes)
+  message            = delete_message.ljust(70, ' ') + attributes_message
+  message_slack(message)
+  logging.info(message)
+
+
+def sync_aws_ad():
+  logger.info('Looking for AWS users to delete that do not exist or are not active in AD')
+  human_aws_accounts      = filter_out_aws_service_accounts(get_all_aws_users())
+  active_ad_users         = get_all_active_ad_users()
+  aws_users_to_be_deleted = []
+
+  for human_aws_account in human_aws_accounts:
+    # users on the whitelist will never be disabled or deleted in anyway.
+    if human_aws_account['UserName'] in aws_users_whitelist:
+      continue
+    if human_aws_account['UserName'] not in active_ad_users:
+      aws_users_to_be_deleted.append(human_aws_account)
+
+  percent_aws_users_deleted = float(len(aws_users_to_be_deleted)) / len(human_aws_accounts)
+  # raise exception if we try to delete too many users as a failsafe.
+  if percent_aws_users_deleted > aws_max_delete_failsafe:
+    exception_msg = 'No users were deleted. %.1f percent to be deleted is beyond the acceptable threshold: %.1f' % (percent_aws_users_deleted * 100, aws_max_delete_failsafe * 100)
+    raise Exception(exception_msg)
+
+  # After the failsafe is over, go through and disable / delete all the users
+  for aws_user_to_be_deleted in aws_users_to_be_deleted:
+    account                  = get_arn_account_profile_name(aws_user_to_be_deleted['Arn'])
+    session                  = boto3.Session(profile_name=account)
+    iam                      = session.client('iam')
+    user_name                = aws_user_to_be_deleted['UserName']
+    user_access_key_metadata = iam.list_access_keys(UserName=user_name)
+    check_if_active_and_disable_user(iam, user_name, user_access_key_metadata, account)
+    if not user_keys_active_recently(iam, user_name, user_access_key_metadata, days_since_active=int(aws_delete_grace_period)):
+      delete_aws_account(iam, user_name, account)
+
+
+if __name__ == '__main__':
+  logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+  logging.getLogger('requests').setLevel(logging.ERROR)
+  logging.getLogger('botocore').setLevel(logging.ERROR)
+  error_counter = 0
+  while True:
+    try:
+      sync_aws_ad()
+    except Exception as error:
+      logger.exception(error)
+      # if we regularly have exceptions, let aws slack know about it once per day.
+      if error_counter == 0:
```

Comment:
error_counter always increments...
if (error_counter%48) == 4   {
   /// only send slack on the 4th successive failure .. notify again  in 24 hours time.
}